### PR TITLE
test: expand TLS SNI edge-case coverage

### LINKED ISSUE #50: test: expand TLS SNI edge-case coverage
## Context
Suggested by code review on #49 (issue #28 fix). The non-UTF-8 SNI handling added there is well-covered for the happy path and the obvious regression, but a few SNI edge cases would benefit from explicit pin-tests so future refactors can't silently change behavior.

## Cases to add
1. **Empty SNI hostname list** — TLS extension present but `list.first()` returns `None`. `extract_sni` should return `None`, no finding emitted, `sni_counts` not touched. (Builder needs to allow zero hostname entries.)
2. **Empty hostname bytes** — `Some((_, b\"\"))`. Currently `std::str::from_utf8(b\"\")` is `Ok(\"\")`, so this counts an empty-string SNI under key `\"\"` and emits no finding. Pin this behavior.
3. **Valid UTF-8 but non-ASCII SNI** (e.g., U-labels with Cyrillic / café) — also a RFC 6066 §3 violation, but currently *not* flagged. Pin the current behavior so the eventual follow-up detection can find it via failing test.
4. **Multi-name SNI list** — only `list.first()` is examined. The current behavior should be pinned so a refactor can't silently change which entry is counted.
5. **Non-UTF-8 SNI when `sni_counts` is at `MAX_MAP_ENTRIES`** — verify the finding still fires even when `sni_counts` is full and the new key gets dropped by `Self::increment`.

## Why not in #49
Each of these requires either expanding `build_client_hello_raw_sni` to handle multiple/empty hostname entries, or constructing a tighter test fixture. Doing it in #49 would scope-creep the fix; doing it as one focused test PR is cleaner.

## Reference
Code review for PR #49 raised these as suggestion S1.

### PR DESCRIPTION
**Pull request by Zious11**
## Summary
- Adds 5 pin-tests for SNI edge cases that the existing non-UTF-8 SNI tests (from #49) don't reach. Each new test documents the *current* analyzer behavior so a future refactor can't silently change it. Closes #50.
- Refactors the test fixture builder to support 0-or-more-entry SNI lists: extracted `build_client_hello_with_sni_list` from `build_client_hello_raw_sni`, preserving the existing API for prior tests.
- One test (#5) intentionally exercises the `MAX_MAP_ENTRIES` boundary by feeding 50,000 unique ClientHellos through the public API. Runtime ~650ms in debug (budget ~2s for CI). No production code changes — this is test-only.
- Code review (pr-test-analyzer) applied pre-merge: RFC 6066 §3 bounds verified, silent-false-positive guard added to test 1, test 5's finding-delta tightened to filter specifically for non-UTF-8 findings, comment drift fixed.
- S4 follow-up (non-zero `NameType`, max-length SNI, trailing bytes) filed as #52.

## New tests
1. **test_sni_extension_with_empty_hostname_list** — `ServerNameList` with zero entries. `extract_sni` returns `None`, no finding, no count, handshake still parses. Guards against tls_parser drift via an explicit `parse_error_count == 0` assertion (empty-list wire form is technically RFC-violating per `server_name_list<1..2^16-1>`; a future tls_parser strictness tightening would otherwise let this test pass for the wrong reason).
2. **test_sni_with_empty_hostname_bytes** — one entry but the hostname bytes are `b""`. Decodes via `from_utf8` as `Ok("")`, counted under the empty-string key, no finding. Degenerate but valid wire form.
3. **test_valid_utf8_non_ascii_sni_currently_not_flagged** — `"café.example"` (a raw U-label, also a RFC 6066 §3 violation per the A-label requirement). Pins current no-finding behavior; the assertion includes the explicit #51 URL so the cross-reference survives grep and will fire loud when that issue lands.
4. **test_multi_name_sni_list_only_first_entry_counted** — 3 ServerName entries; only the first is counted, matching `extract_sni`'s `list.first()` behavior.
5. **test_non_utf8_sni_finding_fires_when_sni_counts_at_capacity** — fills `sni_counts` to `MAX_MAP_ENTRIES = 50_000` via 50k unique ClientHellos through the public API, then sends a non-UTF-8 SNI ClientHello. Verifies the new non-UTF-8 key is dropped at the cap but the finding still fires. Pins a real invariant: finding emission is independent of count insertion. Delta is filtered by `summary.contains("non-UTF-8 bytes")` so the pin stays tight under refactors.

## Why brute-force for test 5 (not a test helper)
Reaching `MAX_MAP_ENTRIES = 50_000` via the public API was chosen over adding a `#[cfg(test)] pub(crate)` test helper on `TlsAnalyzer` to expose internal state. The trade-off was validated with Perplexity (rustc community tends to accept slow boundary tests over test-only production surface) and empirically: debug runtime ~650ms, release ~120ms. The constant is hand-duplicated in the test but self-corrects via the length assertion (any change to the cap would loudly fail).

## RFC validation
Before applying the code-reviewer's fixes, RFC 6066 §3 was fetched directly to confirm the relevant spec claims:
- `ServerNameList server_name_list<1..2^16-1>` — minimum 1 entry (empty lists are RFC-violating wire form, used in test 1 only to exercise the analyzer's defensive branch)
- `enum { host_name(0), (255) } NameType` — `(255)` reserves room for future types (informs the #52 follow-up for non-zero NameType testing)
- RFC 6066 is **silent** on duplicate SNI extensions, so that case was dropped from the follow-up list

## Test plan
- [x] `cargo test --test tls_analyzer_tests` — 18/18 pass (~650ms total, test 5 is ~650ms of that)
- [x] `cargo test` — full suite green
- [x] `cargo clippy --all-targets -- -D warnings` — clean
- [x] `cargo fmt --check` — clean
- [x] Code review pass (`pr-test-analyzer`): 1 important finding (I1 silent false-positive), 1 important (I2 panic message link), 3 suggestions — all applied before PR creation
- [x] RFC 6066 §3 claims validated via direct RFC fetch before applying review fixes

### CODE REVIEW

File: tests/tls_analyzer_tests.rs [hunk 1]
```diff
@@ -20,17 +20,34 @@ fn build_client_hello(sni: &str, cipher_ids: &[u16]) -> Vec<u8> {
...
-    let sni_list_len = (3 + sni_bytes.len()) as u16;
+    // The ServerNameList contents: for each entry, 1 byte host_name type + 2 byte name length + name bytes.
+    let mut sni_list_data = Vec::new();
+    for entry in sni_entries {
+        sni_list_data.push(0x00); // host_name type
+        sni_list_data.extend_from_slice(&(entry.len() as u16).to_be_bytes());
+        sni_list_data.extend_from_slice(entry);
+    }
+    let sni_list_len = sni_list_data.len() as u16;
     let sni_ext_len = 2 + sni_list_len;
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Addressed in d3e7cb3. Switched each length conversion to `u16::try_from(...).expect(...)` with a clear panic message, and `checked_add(2)` for the extension-length calculation. This directly hardens against a scenario in follow-up issue #52 which asks for a 65,534-byte SNI hostname test — with the old `as u16` it would have silently wrapped; now it would panic cleanly if a caller passed something out of range.

File: tests/tls_analyzer_tests.rs [hunk 2]
```diff
@@ -403,3 +420,237 @@ fn test_non_utf8_sni_escapes_control_bytes_in_summary() {
...
+    );
+    // The handshake itself still parsed, so handshake_count should advance.
+    assert_eq!(analyzer.handshake_count(), 1);
+}
+
+#[test]
+fn test_sni_with_empty_hostname_bytes() {
+    // Pin: an SNI hostname of zero bytes (b"") decodes via from_utf8 as Ok(""),
+    // so the analyzer counts it under the empty-string key and emits no finding.
+    // This is a degenerate but technically valid wire form.
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Addressed in d3e7cb3. You are correct — `opaque HostName<1..2^16-1>` makes `b""` RFC-violating (min length 1, not 0). Updated the comment to say "degenerate, RFC-violating" and clarified that the pin is guarding the analyzer's defensive path: tls_parser accepts the non-spec-compliant form anyway, and this test documents the analyzer's response to that degenerate input.

File: tests/tls_analyzer_tests.rs [hunk 3]
```diff
@@ -20,17 +20,42 @@ fn build_client_hello(sni: &str, cipher_ids: &[u16]) -> Vec<u8> {
...
+fn build_client_hello_with_sni_list(sni_entries: &[&[u8]], cipher_ids: &[u16]) -> Vec<u8> {
     let mut extensions = Vec::new();
 
     // SNI extension (type 0x0000)
-    let sni_list_len = (3 + sni_bytes.len()) as u16;
-    let sni_ext_len = 2 + sni_list_len;
+    // The ServerNameList contents: for each entry, 1 byte host_name type + 2 byte name length + name bytes.
+    // Use checked conversions so a future test passing an out-of-range SNI (e.g. the
+    // 65,534-byte hostname requested in issue #52) panics with a clear message rather
+    // than silently wrapping through `as u16` and producing a malformed ClientHello.
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Fixed in 7fbe024 — you are right, each ServerName entry carries 3 bytes of overhead (NameType + u16 length), so the maximum single-entry hostname is u16::MAX − 3 = 65,532 bytes, not 65,534. Comment corrected.

File: tests/tls_analyzer_tests.rs [hunk 4]
```diff
@@ -20,17 +20,42 @@ fn build_client_hello(sni: &str, cipher_ids: &[u16]) -> Vec<u8> {
...
+    let sni_ext_len = sni_list_len
+        .checked_add(2)
+        .expect("SNI extension length would overflow u16");
     extensions.extend_from_slice(&[0x00, 0x00]); // extension type: server_name
     extensions.extend_from_slice(&sni_ext_len.to_be_bytes());
     extensions.extend_from_slice(&sni_list_len.to_be_bytes());
-    extensions.push(0x00); // host_name type
-    extensions.extend_from_slice(&(sni_bytes.len() as u16).to_be_bytes());
-    extensions.extend_from_slice(sni_bytes);
+    extensions.extend_from_slice(&sni_list_data);
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs:60]: Addressed in 7fbe024. Extended the hardening sweep to the three other u16 length fields in `build_client_hello_with_sni_list`:

- `ext_len` (the one you flagged): `u16::try_from(extensions.len()).expect("ClientHello extensions block exceeds u16::MAX bytes")`
- `ciphers_len`: `u16::try_from(ciphers_byte_len).expect("cipher_ids list exceeds u16::MAX bytes (max 32,767 suites)")` — with a `checked_mul(2)` guard on the len*2 arithmetic
- `hs_len` (TLS record body): `u16::try_from(handshake.len()).expect("TLS handshake body exceeds u16::MAX bytes; record fragmentation not implemented")`

Skipped `ch_len` since that one is u24-encoded via bit-shifts, not u16. The 24-bit handshake-length field is harder to overflow (max ~16MB) and the shift form is already defensive since it only uses the low 3 bytes.
